new/property-defaults.md

```diff
@@ -0,0 +1,120 @@
+end
+
+x 'blah' do
+  puts @foo #=> nil
+  puts foo  #=> 1
+  puts foo  #=> 1
+  puts @foo #=> 1
+end
+```
+
+Since this has some utility, we add a new keyword for computed defaults.
+
+#### Computed defaults
+
+```ruby
+class X < Chef::Resource
+  attribute :foo, computed: lazy { Random.rand }
+end
+
+x 'blah' do
+  puts @foo #=> nil
+  puts foo  #=> 1
+  puts foo  #=> 325
+  puts @foo #=> nil
+end
+```
+
+#### Frozen constants
+
+Sticky defaults create a serious issue for constants: *every instance* of a resource gets the same value.
+
+```ruby
+class X < Chef::Resource
+  attribute :foo, default: []
+end
+x 'a' do
+  foo << 1
+  foo << 2
+end
+x 'b' do
+  # ERROR we picked up the values that got added to 'a'
+  puts foo #=> [ 1, 2 ]
+end
+```
+
+To fix this, we propose freezing constant default values, and not assigning them to the instance.
+
+TODO an alternative is to dup them when assigning to an instance the way Chef::Config does.
+
+#### Backcompat break: writing to constant defaults
+
+Users may rely on this behavior to *write* to `myprop`, like this:
+
+```ruby
+class Foo < Chef::Resource
+  attribute :myhash, default: lazy { {} }
+  attribute :mylist, default: lazy { [] }
+end
+foo 'x' do
+  myhash[:a] = 10
+  mylist << 'hi'
+end
+```
+
+This will fail to work in the new world, because we are freezing the constants. This is deliberate and a bugfix; the purpose is to avoid users setting values that affect all instances.
+
+### Validating defaults
+
+We propose that defaults be validated consistently: non-lazy defaults would be validated when the resource class is declared, and lazy defaults would be validated when they are retrieved.
+
+#### Backcompat
+
+Defaults are presently validated inconsistently: they would be validated or not validated based on the order of declaration in the property information. To wit:
+
+```ruby
+# This currently does not validate, even though 10 is not a string
+property :a, kind_of: String, default: 10
+# This currently validates correctly
+property :b, default: 10, kind_of: String
+```
+
+For backcompat purposes, if validation of a default value fails, we will emit a deprecation warning ("you must specify defaults that pass validation").
+
+### Move Core resources to `default` [compatbreak w/deprecation]
+
+`default` is generally a better thing for memory pressure, and is useful for detecting whether a user has set a value or not.  Currently, core Resources tend to assign property default values to the proper instance variable during `initialize`, and it is possible that subclasses may rely on this behavior. For Chef 12, we will grab the default values of any lazy arrays or hashes on initialize, and we will issue a deprecation warning for subclasses that directly use the initialized instance variable.
+
+### `reset_property`
+
+There is already a `property_is_set?`, but no way to reset a property to its default.  We suggest using `reset_property :name` to do this.
```

Comment:
Seems like API creep, this is rare enough that it doesn't seem like a big deal and should basically just work if you do `remove_instance_variable(:@name)` should do the trick. That is slightly less nice, but given I've never wanted to do this or seen anyone want to do it, it seems like a win to avoid adding another method to the API.